File: lab 4_5/Graphs.java
import java.util.*;

// Пример собственного представления дерева (общее дерево, не обязательно бинарное)
class TreeNode {
    String value;
    List<TreeNode> children = new ArrayList<>();

    TreeNode(String value) {
        this.value = value;
    }

    void addChild(TreeNode child) {
        children.add(child);
    }
}

// Пример собственного представления графа (неориентированный граф, список смежности)
class Graph {
    private final Map<String, List<String>> adj = new HashMap<>();

    public void addVertex(String v) {
        adj.putIfAbsent(v, new ArrayList<>());
    }

    public void addEdge(String v1, String v2) {
        addVertex(v1);
        addVertex(v2);
        adj.get(v1).add(v2);
        adj.get(v2).add(v1); // неориентированный граф
    }

    // простой обход в ширину
    public List<String> bfs(String start) {
        List<String> order = new ArrayList<>();
        if (!adj.containsKey(start)) return order;
        Set<String> visited = new HashSet<>();
        Queue<String> queue = new ArrayDeque<>();
        visited.add(start);
        queue.add(start);
        while (!queue.isEmpty()) {
            String v = queue.remove();
            order.add(v);
            for (String neighbor : adj.get(v)) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.add(neighbor);
                }
            }
        }
        return order;
    }
}

public class Main {
    // печать дерева с отступами
    public static void printTree(TreeNode node, int level) {
        if (node == null) return;
        for (int i = 0; i < level; i++) {
            System.out.print("  ");
        }
        System.out.println(node.value);
        for (TreeNode child : node.children) {
            printTree(child, level + 1);
        }
    }

    public static void main(String[] args) {
        // ---------- Пример дерева: структура меню веб-сайта ----------
        TreeNode root = new TreeNode("Главная страница");
        TreeNode products = new TreeNode("Каталог товаров");
        TreeNode about = new TreeNode("О компании");
        TreeNode contact = new TreeNode("Контакты");

        root.addChild(products);
        root.addChild(about);
        root.addChild(contact);

        TreeNode electronics = new TreeNode("Электроника");
        TreeNode clothes = new TreeNode("Одежда");
        products.addChild(electronics);
        products.addChild(clothes);

        System.out.println("Структура меню (дерево):");
        printTree(root, 0);

        // ---------- Пример графа: дружеские связи пользователей ----------
        Graph social = new Graph();
        social.addEdge("Антон", "Борис");
        social.addEdge("Антон", "Дима");
        social.addEdge("Борис", "Егор");
        social.addEdge("Дима", "Егор");
        social.addEdge("Егор", "Инна");

        System.out.println();
        System.out.println("Обход графа дружбы из вершины 'Антон':");
        System.out.println(social.bfs("Антон"));
    }
}

File: lab 4_5/Graphs.py
from collections import deque

# Собственное представление общего дерева (каждый узел может иметь много детей)
class TreeNode:
    def __init__(self, value: str) -> None:
        self.value = value
        self.children: list["TreeNode"] = []

    def add_child(self, child: "TreeNode") -> None:
        self.children.append(child)


def print_tree(root: TreeNode, level: int = 0) -> None:
    indent = "  " * level
    print(f"{indent}{root.value}")
    for child in root.children:
        print_tree(child, level + 1)


# Собственное представление неориентированного графа через список смежности
class Graph:
    def __init__(self) -> None:
        self.adj: dict[str, list[str]] = {}

    def add_vertex(self, v: str) -> None:
        if v not in self.adj:
            self.adj[v] = []

    def add_edge(self, v1: str, v2: str) -> None:
        # добавляем вершины, если их ещё нет
        self.add_vertex(v1)
        self.add_vertex(v2)
        # так как граф неориентированный, добавляем ребро в обе стороны
        self.adj[v1].append(v2)
        self.adj[v2].append(v1)

    def bfs(self, start: str) -> list[str]:
        """Простой обход в ширину от вершины start."""
        if start not in self.adj:
            return []
        visited: set[str] = {start}
        queue: deque[str] = deque([start])
        order: list[str] = []
        while queue:
            v = queue.popleft()
            order.append(v)
            for neighbor in self.adj[v]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        return order


if __name__ == "__main__":
    # -------- Пример дерева: структура категорий интернет-магазина --------
    root = TreeNode("Все категории")
    electronics = TreeNode("Электроника")
    clothes = TreeNode("Одежда")
    home = TreeNode("Товары для дома")

    root.add_child(electronics)
    root.add_child(clothes)
    root.add_child(home)

    phones = TreeNode("Смартфоны")
    laptops = TreeNode("Ноутбуки")
    electronics.add_child(phones)
    electronics.add_child(laptops)

    print("Структура категорий (дерево):")
    print_tree(root)

    # -------- Пример графа: сеть городов и дорог между ними --------
    g = Graph()
    g.add_edge("Москва", "Тула")
    g.add_edge("Москва", "Казань")
    g.add_edge("Тула", "Воронеж")
    g.add_edge("Казань", "Нижний Новгород")
    g.add_edge("Воронеж", "Ростов-на-Дону")

    print("\nОбход графа дорог из города 'Москва':")
    print(g.bfs("Москва"))